packages/coreutils/src/poll.ts

```diff
@@ -0,0 +1,657 @@
+   *
+   * @returns A promise that resolves after tick is scheduled and never rejects.
+   */
+  async start(): Promise<void> {
+    if (this.isDisposed) {
+      return;
+    }
+
+    if (this.state.phase === 'constructed') {
+      await this.ready;
+    }
+
+    if (this.state.phase === 'standby' || this.state.phase === 'stopped') {
+      this.schedule({
+        interval: Private.IMMEDIATE,
+        payload: null,
+        phase: 'started',
+        timestamp: new Date().getTime()
+      });
+    }
+  }
+
+  /**
+   * Stops the poll. Schedules `stopped` tick if necessary.
+   *
+   * @returns A promise that resolves after tick is scheduled and never rejects.
+   */
+  async stop(): Promise<void> {
+    if (this.isDisposed) {
+      return;
+    }
+
+    if (this.state.phase === 'constructed') {
+      await this.ready;
+    }
+
+    if (this.state.phase !== 'stopped') {
+      this.schedule({
+        interval: Private.NEVER,
+        payload: null,
+        phase: 'stopped',
+        timestamp: new Date().getTime()
+      });
+    }
+  }
+
+  /**
+   * A promise that resolves after the poll has scheduled its first tick.
+   *
+   * #### Notes
+   * This accessor is protected to allow sub-classes to implement methods that
+   * can `await this.ready` if `this.state.phase === 'constructed'`.
+   *
+   * A poll should handle ready state without needing to expose it to clients.
+   */
+  protected readonly ready: Promise<void>;
+
+  /**
+   * Schedule the next poll tick.
+   *
+   * @param tick - The new tick data to populate the poll state.
+   *
+   * #### Notes
+   * This method is protected to allow sub-classes to implement methods that can
+   * schedule poll ticks.
+   *
+   * This method synchronously modifies poll state so it should be invoked with
+   * consideration given to the poll state that will be overwritten. It will
+   * typically be invoked in a method that returns a promise, allowing the
+   * caller e.g. to `await this.ready` before scheduling a new tick.
+   */
+  protected schedule(tick: IPoll.Tick<T, U>): void {
+    const current = new PromiseDelegate<this>();
+    const pending = this._tick;
+    const execute = () => {
+      if (this.isDisposed || this.tick !== current.promise) {
+        return;
+      }
+
+      this._execute();
+    };
+
+    // Clear the schedule if possible.
+    if (this.state.interval === Private.IMMEDIATE) {
+      cancelAnimationFrame(this._timeout);
+    } else {
+      clearTimeout(this._timeout);
+    }
+
+    // Update poll state and schedule the next tick.
+    this._state = tick;
+    this._tick = current;
+    this._timeout =
+      tick.interval === Private.IMMEDIATE
+        ? requestAnimationFrame(execute)
+        : tick.interval === Private.NEVER
+          ? -1
+          : setTimeout(execute, tick.interval);
+
+    // Resolve the pending poll promise and emit the ticked signal.
+    void pending.promise.then(() => {
+      this._ticked.emit(tick);
```

Comment:
Why are we emitting in the promise? It seems like it would be better to emit synchronous with setting `this._state` and `this._tick`, so you get notified anytime the state of the poll changes. Emitting in the promise means that we'll sometime later get notified that sometime in the past the state of the poll changed, and it may not even be the current state anymore.

Reply:
All scheduling happens asynchronously, so emitting the promise and the signal in this order vs. emitting the signal first then resolving the promise should not make a difference, the emission happens *before* the state can get replaced. If you think it's clearer to have the emission happen before the promise resolution, I don't have any objection. I don't think there's a potential for a race condition here, though. Do you?

Reply:
Well, scheduling happens synchronously, but the timeouts fire asynchronously, of course. I think the current behavior of waiting until the pending promise resolves (i.e., another frame) to fire the tick emission allows the tick emission to be out of date. I think it's clearer to fire the tick emission when scheduled, when the state is updated, and before the tick promise is resolved. For example, since we schedule the requestAnimationFrame before resolving the pending promise, it's possible that the execute function for the next time around will run before the pending promise resolves. In fact, it's probably better to resolve that pending promise before doing the setTimeout.

Also, firing the signal synchronous means that your slot can inspect the state of the object, and the tick is consistent with the poll state.

Reply:
Upon reflection, I agree and pushed an update.